src/gen_multisig/gen_multisig.cpp

```diff
@@ -0,0 +1,235 @@
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+// 
+// Parts of this file are originally copyright (c) 2012-2013 The Cryptonote developers
+
+/*!
+ * \file gen_multisig.cpp
+ * 
+ * \brief Generates a set of multisig wallets
+ */
+#include <iostream>
+#include <sstream>
+#include <boost/program_options.hpp>
+#include <boost/algorithm/string.hpp>
+#include <boost/format.hpp>
+#include "include_base_utils.h"
+#include "crypto/crypto.h"  // for crypto::secret_key definition
+#include "common/i18n.h"
+#include "common/command_line.h"
+#include "common/util.h"
+#include "common/scoped_message_writer.h"
+#include "wallet/wallet_args.h"
+#include "wallet/wallet2.h"
+
+using namespace std;
+using namespace epee;
+using namespace cryptonote;
+using boost::lexical_cast;
+namespace po = boost::program_options;
+
+#undef MONERO_DEFAULT_LOG_CATEGORY
+#define MONERO_DEFAULT_LOG_CATEGORY "wallet.gen_multisig"
+
+namespace genms
+{
+  const char* tr(const char* str)
+  {
+    return i18n_translate(str, "tools::gen_multisig");
+  }
+
+}
+
+namespace
+{
+  const command_line::arg_descriptor<std::string> arg_filename_base = {"filename-base", genms::tr("Base filename (-1, -2, etc suffixes will be appended as needed)"), ""};
+  const command_line::arg_descriptor<std::string> arg_scheme = {"scheme", genms::tr("Give threshold and participants at once as M/N"), ""};
+  const command_line::arg_descriptor<uint32_t> arg_participants = {"participants", genms::tr("How many participants wil share parts of the multisig wallet"), 0};
+  const command_line::arg_descriptor<uint32_t> arg_threshold = {"threshold", genms::tr("How many signers are required to sign a valid transaction"), 0};
+  const command_line::arg_descriptor<bool, false> arg_testnet = {"testnet", genms::tr("Create testnet multisig wallets"), false};
+
+  const command_line::arg_descriptor< std::vector<std::string> > arg_command = {"command", ""};
+}
+
+static bool generate_multisig(uint32_t threshold, uint32_t total, const std::string &basename, bool testnet)
+{
+  tools::msg_writer() << (boost::format(genms::tr("Generating %u %u/%u multisig wallets")) % total % threshold % total).str();
+
+  const auto pwd_container = tools::wallet2::password_prompt(true);
+
+  try
+  {
+    // create M wallets first
+    std::vector<boost::shared_ptr<tools::wallet2>> wallets(total);
+    for (size_t n = 0; n < total; ++n)
+    {
+      std::string name = basename + "-" + std::to_string(n + 1);
+      wallets[n].reset(new tools::wallet2(testnet));
+      wallets[n]->init("");
+      wallets[n]->generate(name, pwd_container->password(), rct::rct2sk(rct::skGen()), false, false);
+    }
+
+    // gather the keys
+    std::vector<crypto::secret_key> sk(total);
+    std::vector<crypto::public_key> pk(total);
+    for (size_t n = 0; n < total; ++n)
+    {
+      if (!wallets[n]->verify_multisig_info(wallets[n]->get_multisig_info(), sk[n], pk[n]))
```

Comment:
You may call me pedantic, but I'd like static functions to be called without referring to an instance: `tools::wallet2:: verify_multisig_info`, likewise for the `_extra_` version below.